File: Java/pattern/ObserverProj/src/com/pattern/ConcreteSubject.java
package com.pattern;

import java.util.ArrayList;

public class ConcreteSubject implements Subject {
	private ArrayList<Observer> collect = new ArrayList<Observer>();
	
	public void addObserver(Observer ob) {
		if (!collect.contains(ob))
			collect.add(ob);
	}
	
	public void deleteObserver(Observer ob) {
		if (collect.contains(ob))
			collect.remove(ob);
	}

	@Override
	public void nofityAll(String msg) {
		int collectSize = collect.size();
		for (int i = 0; i < collectSize; ++i)
			collect.get(i).update(msg);
	}

}

File: cplusplus/design_pattern/ObserverProj/ConcreteSubject.cpp
#include "stdafx.h"
#include "ConcreteSubject.h"


ConcreteSubject::ConcreteSubject()
{
}


ConcreteSubject::~ConcreteSubject()
{
}


void ConcreteSubject::notifyAll(std::string msg)
{
    std::for_each(collect.begin(), collect.end(), [msg](auto it) { it->update(msg); });
}

void ConcreteSubject::addObserver(std::shared_ptr<Observer> ob)
{
    if (!(std::find(collect.begin(), collect.end(), ob) != collect.end())) {
        collect.push_back(ob);
    }
}


void ConcreteSubject::deleteObserver(std::shared_ptr<Observer> ob)
{
    auto it = collect.begin();
    if ((it = std::find(collect.begin(), collect.end(), ob)) != collect.end()) {
        collect.erase(it);
    }
}
